p2p/peerexchange/discovery.go

```diff
@@ -75,32 +76,59 @@ func New(logger log.Log, h host.Host, config Config) (*Discovery, error) {
 		}
 		bootnodes = append(bootnodes, info)
 	}
+
+	var advertise ma.Multiaddr
+	if len(config.AdvertiseAddress) > 0 {
+		var err error
+		advertise, err = ma.NewMultiaddr(config.AdvertiseAddress)
+		if err != nil {
+			return nil, fmt.Errorf("address to advertise (%s) is invalid: %w", config.AdvertiseAddress, err)
+		}
+		for _, proto := range advertise.Protocols() {
+			if proto.Code == ma.P_P2P {
+				return nil, fmt.Errorf("address to advertise (%s) includes p2p identity", advertise.String())
+			}
+		}
+	} else {
+		var err error
+		advertise, err = ma.NewComponent("tcp", strconv.Itoa(int(portFromHost(logger, h))))
+		if err != nil {
+			return nil, fmt.Errorf("create tcp multiaddr %w", err)
+		}
+	}
 	best, err := bestHostAddress(h)
 	if err != nil {
 		return nil, err
 	}
 	d.book.AddAddresses(bootnodes, best)
 
-	protocol := newPeerExchange(h, d.book, portFromHost(logger, h), logger, config.PeerExchange)
+	protocol := newPeerExchange(h, d.book, advertise, logger, config.PeerExchange)
 	d.crawl = newCrawler(h, d.book, protocol, logger)
 	sub, err := h.EventBus().Subscribe(new(event.EvtLocalAddressesUpdated), eventbus.BufSize(4))
 	if err != nil {
 		return nil, fmt.Errorf("failed to subscribe to eventbus: %w", err)
 	}
-	d.eg.Go(func() error {
-		defer sub.Close()
-		for {
-			select {
-			case <-sub.Out():
-				port := portFromHost(logger, h)
-				if port != 0 {
-					protocol.UpdateExternalPort(port)
+	if len(config.AdvertiseAddress) == 0 {
+		d.eg.Go(func() error {
+			defer sub.Close()
+			for {
+				select {
+				case <-sub.Out():
+					port := portFromHost(logger, h)
+					if port != 0 {
+						advertise, err := ma.NewComponent("tcp", strconv.Itoa(int(portFromHost(logger, h))))
+						if err != nil {
+							logger.With().Error("failed to create tcp multiaddr", log.Err(err))
+						} else {
+							protocol.UpdateAdvertisedAddress(advertise)
+						}
+					}
```

Comment:
can advertise from line 94 reused here?

```suggestion
				case <-sub.Out():
					protocol.UpdateAdvertisedAddress(advertise)
```

it has already been checked to be valid as well.